x/qgb/orchestrator/orchestrator.go

```diff
@@ -4,95 +4,436 @@ import (
+	if err != nil {
+		return err
 	}
+	orch.logger.Info("signed Valset", "nonce", msg.Nonce, "tx_hash", hash)
 	return nil
 }
+
+func (orch Orchestrator) processDataCommitmentEvent(
+	ctx context.Context,
+	dc types.DataCommitment,
+) error {
+	commitment, err := orch.querier.QueryCommitment(
+		ctx,
+		fmt.Sprintf("block.height >= %d AND block.height <= %d",
+			dc.BeginBlock,
+			dc.EndBlock,
+		),
+	)
+	if err != nil {
+		return err
+	}
+	dataRootHash := types.DataCommitmentTupleRootSignBytes(types.BridgeId, big.NewInt(int64(dc.Nonce)), commitment)
+	dcSig, err := types.NewEthereumSignature(dataRootHash.Bytes(), &orch.evmPrivateKey)
+	if err != nil {
+		return err
+	}
+
+	msg := types.NewMsgDataCommitmentConfirm(
+		commitment.String(),
+		ethcmn.Bytes2Hex(dcSig),
+		orch.signer.GetSignerInfo().GetAddress(),
+		orch.orchEthAddress,
+		dc.BeginBlock,
+		dc.EndBlock,
+		dc.Nonce,
+	)
+	hash, err := orch.broadcaster.BroadcastTx(ctx, msg)
+	if err != nil {
+		return err
+	}
+	orch.logger.Info("signed commitment", "nonce", msg.Nonce, "begin_block", msg.BeginBlock, "end_block", msg.EndBlock, "commitment", commitment, "tx_hash", hash)
+	return nil
+}
+
+var _ Broadcaster = &broadcaster{}
+
+type Broadcaster interface {
+	BroadcastTx(ctx context.Context, msg sdk.Msg) (string, error)
+}
+
+type broadcaster struct {
+	mutex   *sync.Mutex
+	signer  *paytypes.KeyringSigner
+	qgbGrpc *grpc.ClientConn
+}
+
+func NewBroadcaster(qgbGrpcAddr string, signer *paytypes.KeyringSigner) (Broadcaster, error) {
+	qgbGrpc, err := grpc.Dial(qgbGrpcAddr, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	return &broadcaster{
+		mutex:   &sync.Mutex{}, // investigate if this is needed
+		signer:  signer,
+		qgbGrpc: qgbGrpc,
+	}, nil
+}
+
+func (bc *broadcaster) BroadcastTx(ctx context.Context, msg sdk.Msg) (string, error) {
+	bc.mutex.Lock()
+	defer bc.mutex.Unlock()
+	err := bc.signer.QueryAccountNumber(ctx, bc.qgbGrpc)
+	if err != nil {
+		return "", err
+	}
+
+	builder := bc.signer.NewTxBuilder()
+	// TODO make gas limit configurable
+	builder.SetGasLimit(9999999999999)
+	// TODO: update this api
+	// via https://github.com/celestiaorg/celestia-app/pull/187/commits/37f96d9af30011736a3e6048bbb35bad6f5b795c
+	tx, err := bc.signer.BuildSignedTx(builder, msg)
+	if err != nil {
+		return "", err
+	}
+
+	rawTx, err := bc.signer.EncodeTx(tx)
+	if err != nil {
+		return "", err
+	}
+
+	// FIXME sdktypestx.BroadcastMode_BROADCAST_MODE_BLOCK waits for a block to be minted containing
+	// the transaction to continue. This makes the orchestrator slow to catchup.
+	// It would be better to just send the transaction. Then, another job would keep an eye
+	// if the transaction was included. If not, retry it. But this would mean we should increment ourselves
+	// the sequence number after each broadcasted transaction.
+	// We can also use BroadcastMode_BROADCAST_MODE_SYNC but it will also fail due to a non incremented
+	// sequence number.
```

Comment:
@evan-forbes please when you have time,  let me know what you think about this,

Reply:
we might want to create larger txs with many sdk.Messages and still use broadcast mode sync just for simplicity, but I can definitely see where a more async approach would be desirable

Reply:
Should we create an issue for that?